feed.xml

```diff
@@ -14,6 +14,436 @@ Nick Ripley's blog. Programming and other interests.
 <docs>https://www.rssboard.org/rss-specification</docs>
 <atom:link href="https://www.nsrip.com/feed.xml" rel="self" type="application/rss+xml"/>
 
+<item>
+<title>
+Don't clobber the frame pointer
+</title>
+<link>
+https://www.nsrip.com/posts/clobberfp.html
+</link>
+<description>
+Recently I diagnosed and fixed two frame pointer unwinding crashes in Go.
+The root causes were two flavors of the same problem:
+buggy assembly code clobbered a frame pointer.
+By "clobbered" I mean wrote over the value without saving &amp; restoring it.
+One bug clobberd the frame pointer register.
+The other bug clobbered a frame pointer saved on the stack.
+This post explains the bugs,
+talks a bit about ABIs and calling conventions,
+and makes some recommendations for how to avoid the bugs.
+</description>
+<pubDate>Tue, 24 Dec 2024 10:30:00 EDT</pubDate>
+<guid>
+https://www.nsrip.com/posts/clobberfp.html
+</guid>
+<content:encoded>
+<![CDATA[
+<title >Don't clobber the frame pointer</title>
+<h1 >Don't clobber the frame pointer</h1>
+
+<p ><b >
+<a href="https://nsrip.com/posts/../index.html">Home</a>
+-
+<a href="https://nsrip.com/posts/../feed.xml">RSS</a>
+</b></p>
+
+<p ><i >Posted 2024-12-24</i></p>
+
+<p >
+Recently I diagnosed and fixed two frame pointer unwinding crashes in Go.
+The root causes were two flavors of the same problem:
+buggy assembly code clobbered a frame pointer.
+By "clobbered" I mean wrote over the value without saving &amp; restoring it.
+One bug clobbered the frame pointer register.
+The other bug clobbered a frame pointer saved on the stack.
+This post explains the bugs,
+talks a bit about ABIs and calling conventions,
+and makes some recommendations for how to avoid the bugs.
+</p>
+
+<p >
+Here's the short version of what you should do when writing assembly for Go to avoid the problems discussed in this post:
+<ul >
+<li >
+First, read the <a href="https://go.dev/doc/asm">Go assembly guide</a>.
+</li>
+<li >
+Prefer an assembly generator like <a href="https://github.com/mmcloughlin/avo" target="_blank">Avo</a>.
+Tools like Avo are aware of the underlying rules for using registers and manipulating the stack,
+and generally make writing non-trivial amounts of assembly easier.
+</li>
+<li >
+If your assembly function calls Go functions,
+prefer not to use the frame pointer registers at all
+(<code >BP</code> for AMD64, <code >R29</code> for ARM64).
+Leaving a non-frame pointer value in those registers prior to calling a Go function can crash the runtime execution tracer and profilers.
+</li>
+<li >
+Otherwise, if your function has a non-zero frame size,
+the assembler will correctly make a stack frame for you,
+growing the stack if needed,
+and properly save and restore the frame pointer.
+The means the frame pointer will be safe to use.
+Consider giving your function a small, non-zero frame size even if you don't <i >need</i> to if you want to use the frame pointer.
```

Comment:
To make sure I follow: The reason this is safe is because you're not expecting the ASM function to be interrupted for frame pointer unwinding? Does the runtime not do async preemption for ASM functions?

Also: This still ends up breaking external frame pointer unwinders such as linux perf, right?

Reply:
Right, the runtime doesn't preempt assembly functions: https://cs.opensource.google/go/go/+/master:src/runtime/preempt.go;l=407-416;drc=2c635b68fdc8ddf83208ed2ec65eff09a3af58b8. It's not totally safe to use the frame pointer even then, though, if the assembly code calls back into regular Go functions. No idea if that's commonly done. And yeah, it'll break unwinding for perf if you put something else in the frame pointer. Though at least in that case it's still "safe" in the sense that your program won't crash. Honestly, "don't touch the frame pointer at all" might be better advice :D